### ISSUE TITLE
add export / import support for Fountain files

### ISSUE THREAD

**Issue by Reaper10** [jgm/pandoc#4228]
add export / import support for Fountain files
https://fountain.io/
https://fountain.io/syntax
https://github.com/nyousefi/Fountain

**Comment by acrisci** [comment]
Since Fountain is for screenplays, there's a lot of specialized elements for things like scene transitions and character dialogue that aren't going to be defined in a normal document format. And in general, you can't interpret an arbitrary document as a screenplay. So I believe it is not possible to create a Fountain writer.

However these elements do have a fairly standard format representation on the page. For instance, dialogue is center aligned and transitions are right aligned. So a reader would be possible, although you would lose the intent of the formatting.

But the reader is really the most useful thing because the most common use case for this will be to format a screenplay written in Fountain to something printable for the actors.

I'm going to take a look at this and see if it's something I could possibly do.

**Comment by acrisci** [comment]
Now I don't think Fountain would be a good fit for pandoc. I recommend closing this as out of scope.

Fountain is not a normal document so its structure cannot be expressed in pandoc form. Conversion would be primarily concerned with presentation. For instance, pandoc has no way to express right-aligned blocks which would be necessary. But there are also some other really complicated presentational aspects of Fountain such as simultaneous dialogue which I think are almost certainly out of scope.

Right aligned blocks were stated to be out of scope in #4451

> Pandoc tries to capture structure, not presentational things like this.

**Comment by alerque** [comment]
I'm not quite convinced. I do see the counter argument, but there are actually quite a few formats for which reasonable representations of intent could be produced, and even a few that could be round-tripped. It's a bit out of the edge I admit, but I'm not convinced it's completely out of scope. It's not a presentation format, it's a specialized take on a plain text format in the same way Markdown is a generalized take on a plain text format. Many of the formats already supported have specialized elements. The question is how many of them can be mapped in a way that would be reasonably useful?

**Comment by acrisci** [comment]
> The question is how many of them can be mapped in a way that would be reasonably useful?

With presentation out of the picture, there is not a good mapping to the pandoc format. Some things map well. Scene headings are headings. Actions are paragraphs. But what is a scene transition if you can't say it's a "right aligned block" and what is dialogue if you can't say it's a "center aligned block"? Maybe that could be solved but then you have simultaneous dialogue which is roughly "a grid block of dialogue blocks" which means you'd have to take into consideration nested blocks which is not something a general document format supports.

For reference, see this [image](https://fountain.io/_img/fountain-quickreference.jpg) of what conversion would look like.

The alternative is to parse the screenplay and have elements for these things and then put those into the pandoc document format. That represents a pretty big scope increase and it's not at all clear how a markdown writer would handle something like a "scene transition" since there's not anything like that in the markdown spec. And other formats like html could only write that out in terms of presentation. No other document format I know has first class support for sceenplay elements.

**Comment by jgm** [comment]
Pandoc has Divs with attributes.  A dialogue could go into a Div with the `attribute` class.  Then it would be possible to format this as a centered block.
A scene transition could be a specially marked Div.
Simultaneous dialogue could be a specially marked Div containing two Divs.
I see no reason to think that the pandoc AST couldn't capture the *structure* of a Fountain screenplay.  Rendering is another issue; proper rendering might require a filter to interpret the special Divs for the target format, but that's entirely possible.

**Comment by rubensz** [comment]
Thank you @acrisci for lifting the technical challenges with simultaneous dialogue and other elements. In my experience, simultaneous dialogue is rarely used.  The technical solutions needed for using pandoc and fountain is out of my depth. But I know that having pandoc come to the rescue for us screenwriters would be a true blessing. There is Afterwriting but it's a bit clunky with downloading the files to a directory, not offence to the developer. Would be so amazing to just have CLI app and thats it. If anybody with the skillset could take a look at this, at least this writer would be forever grateful.

**Comment by jgm** [comment]
Would the ability to write fountain or the ability to read it be more useful?
Looks like there are a lot of apps available to process fountain: https://fountain.io/apps
Just trying to get a better sense of why pandoc would be useful.

**Comment by rubensz** [comment]
Thank you for considering this feature, @jgm! 

My guess is that 99.9% of use cases would be that you have your screenplay that you have written in plain text, using Fountain, but if you want to share it with a director or producer, you need to have it exported to the industry standard format and share it as a PDF. 

There are many proprietary application that does this but they are bloaty and expensive. The closets things to a CLI-program for exporting fountain-screenplays is [Afterwriting](https://github.com/ifrost/afterwriting-labs/blob/master/docs/clients.md) which both has an online-application and a CLI which uses node.js. But you have to download the files to the working directory and it can be a little bit clunky. Would be amazing if pandoc could handle this, perhaps with a handful of flags for some options. 

I don't have any relevant skills make a fork myself and contribute but would be forever grateful to whoever would be interested in this.

**Comment by acrisci** [comment]
I think that the approach in this issue is a good one. I just started working on something else and forgot about this.

**Comment by jgm** [comment]
I don't think a fountain reader would be to hard to create.
There'd be questions about what it should parse things to -- one could use Divs and Spans with classes, but then a lua filter would be needed to format appropriately in PDF/LaTeX.

**Comment by earthbound19** [comment]
[wrap](https://github.com/Wraparound/wrap) is a CLI tool that converts fountain format to pdf. Its parsing of fountain might inform any effort on this.

IMO dual dialogue in screenplays should not exist, and I would not support it if I wrote a converter. 

Why it's a bad idea:
- It's a programming and layout nightmare to handle
- It only addresses the case for two characters, it doesn't address the case of many characters speaking at the same time. What do you do with three characters who speak at the same time: dual dialogue followed by action that mentions that a third character speaks at the same time? For four characters? Dual dialogue followed by action that says two other characters also speak at the same time followed by dual dialogue? Too complex. What you should do with three, four or more characters is what you should do with two: just mention in action that they all speak at the same time, then write it out for each of them under character slugs as usual. Or just write in action that they all say some things of some nature at the same time, and leave it to the actors to improvise.
- Characters speaking at the same time can be improvisational territory that doesn't need to be written verbatim and followed besides

**Comment by rubensz** [comment]
@earthbound19 Thanks! Looks cool, will check it out!

**Comment by AndreaMonzini** [comment]
Hello, i add some info that could be useful to use Fountain with Pandoc.
Personally i think that would be useful to read Fountain with Pandoc and be able to export it ( HTML, EPUB, PDF, etc...)

The Courier Prime font optimized for screenplays with Open Font License (OFL) 
https://quoteunquoteapps.com/courierprime/index.php

I contributed to a Fountain Blender addon by @tin2tin
https://github.com/tin2tin/Blender_Screenwriter

Fountain to PDF by @Tagirijus
https://github.com/Tagirijus/fountain2pdf

screenplain by@vilcans
https://github.com/vilcans/screenplain/blob/master/screenplain/export/pdf.py
https://github.com/vilcans/screenplain

thank you

**Comment by rubensz** [comment]
Great tips, @AndreaMonzini! Thanks! 
I just started this repo called awesome screenwriting tools. 
Want to be the first contributor? 
https://github.com/rubensz/awesome-screenwriting-tools

**Comment by AndreaMonzini** [comment]
Great list @rubensz ! 
I hope that  Pandoc will be added to command line tools of course !

**Comment by paulrpotts** [comment]
Hello, came here via a Google search to see if pandoc supported Fountain... just came across Fountain today.

I'm not an actual screenwriter, just a blogger/newsletter writer, but I use Markdown & pandoc every week to produce HTML for a Tinyletter newsletter, and have been experimenting for a while with other output formats to eventually let me organize posts into books.

My primary interest would be reading Fountain, not writing it. I get that the distinction between semantic parts and layout is even worse than normal with this format - definitely not ideal, but I'd still like to see it working as I hope to get into screenplay formats, particularly for writing radio drama. So a nice way to generate an HTML fragment from a Fountain file using pandoc would be great for me. I don't care about the simultaneous dialogue thing - that seems ugly and not really required. Way back when I wrote short stories that played with typography and featured simultaneous narrations in parallel along the page, but for that I would likely turn to a real WYSIWYG page layout tool of some kind.

**Comment by tucnak** [comment]
I subscribe to everything @AndreaMonzini is saying. Being a screenwriter myself, I've come to appreciate simplicity of Fountain as a thing of its own more and more over the last couple days. "Dual dialog" style of writing is complete nonsence, and I don't think anyone in the industry does this. I mean, you can come up with whatever wicked form of script, but in the end the actors expect to read a certain thing which is more or less uniform w.r.t. the common script structure. I wish I was sufficiently proficient in Haskell to bring this home; that said, I don't see a single reason why pandoc wouldn't be able to support Fountain. Having successfully used it for Latex, Markdown and the numerous likes of it, I'd expect it to support Fountain as well, to whatever extent possible.

**Comment by mb21** [comment]
Since [wrap](https://github.com/Wraparound/wrap) is open source and already converts Fountain to HTML and PDF... (presumably with better support for some of the more advanced fountain features than we could ever implement in pandoc and its fairly generic document AST)... anything in particular people would want from pandoc? Is it just because you have pandoc already installed, or..?

**Comment by tucnak** [comment]
@mb21 I, for one, don't like certain aspects of Fountain and would rather generate AST from a slight dialect of it, suitable for different exotic scenarios, like AI inserts (generated text) and stuff like that. This would be trivial to do if pandoc implemented a Fountain writer.

**Comment by jgm** [comment]
Have you thought about creating a custom lua writer for fountain?
https://pandoc.org/MANUAL.html#custom-writers

**Comment by jgm** [comment]
Pandoc now allows the use of custom readers, written in Lua.  I have written a reader for Fountain -- it probably still needs a lot of work, but it's probably a good start.

I'm attaching the script here, as well as a reference-doc designed to be used with it if you want to produce a docx from your fountain script.  (For other formats, more customization would be needed -- CSS in the case of HTML, a Lua filter for LaTeX/PDF.)

To use it,
```
pandoc -f fountain.lua sample.fountain  -o sample.docx --reference-doc fountain-ref.docx
```

[fountain-ref.docx](https://github.com/jgm/pandoc/files/7611020/fountain-ref.docx)
[fountain.lua.txt](https://github.com/jgm/pandoc/files/7611022/fountain.lua.txt) (you'll need to rename this as `fountain.lua`)

**Comment by jgm** [comment]
I have now set up a little repository containing the custom reader and the necessary auxiliary files, plus instructions for converting Fountain files to HTML, EPUB, PDF, and Docx using pandoc.
https://github.com/pandoc/pandoc-fountain

**Comment by iwelch** [comment]
fountain to pdf:

```
$ brew install wkhtmltopdf
Warning: wkhtmltopdf has been deprecated because it is discontinued upstream!
```
